### PR TITLE
Configuration

### PR DESCRIPTION
**Pull request by jajik**
Mainly add an initial support for attributes.

### CODE REVIEW

File: softwareComponents/configuration/include/configuration/serialization.hpp [hunk 1]
```diff
@@ -31,14 +31,14 @@ namespace rofi::configuration::serialization {
...
-    template< typename M >
-    M fromJSON( const nlohmann::json& j, ModuleId id );
+    template< typename M > requires std::derived_from< M, Module >
+    M moduleFromJSON( const nlohmann::json& j, ModuleId id );
 
-    template< typename M >
-    M fromJSON( const nlohmann::json& j ) {
+    template< typename M > requires std::derived_from< M, Module >
+    M moduleFromJSON( const nlohmann::json& j ) {
         assert( j.is_object() && ++j.begin() == j.end() && "json is not a single object" );
```
Comment (Ondraceq) [inline softwareComponents/configuration/include/configuration/serialization.hpp:39]: This is undefined when there is no object.
Also I would like it to throw an exception, since it's essentialy reading the user input (so you want to catch that even in production code).

PS. And `stoi` can already throw.

File: softwareComponents/configuration/include/configuration/serialization.hpp [hunk 2]
```diff
@@ -89,19 +106,21 @@ namespace rofi::configuration::serialization {
...
         j[ "beta"  ] = m.getBeta().deg();
         j[ "gamma" ] = m.getGamma().deg();
 
+        addAttributes( j, attrCb, m );
         return { { std::to_string( m.getId() ), j } };
     }
 
     template<>
-    inline UniversalModule fromJSON( const nlohmann::json& j, ModuleId id ) {
+    inline UniversalModule moduleFromJSON( const nlohmann::json& j, ModuleId id ) {
```
Comment (Ondraceq) [inline softwareComponents/configuration/include/configuration/serialization.hpp:123]: Consider adding the specification type explicitly `inline UniversalModule moduleFromJSON<UniversalModule>(`. I'm not 100% sure it's much better, since you repeat yourself, but it makes clear for me that it's a specialization.

File: softwareComponents/configuration/include/configuration/serialization.hpp [hunk 3]
```diff
@@ -250,26 +277,30 @@ namespace rofi::configuration::serialization {
...
 
         return res;
     }
 
-    template<>
+    inline nlohmann::json toJSON( const Rofibot& bot ) {
+        return toJSON( bot, [](auto ...){ return nlohmann::json{}; } );
+    }
+
     inline Rofibot fromJSON( const nlohmann::json& j ) {
```
Comment (Ondraceq) [inline softwareComponents/configuration/include/configuration/serialization.hpp:291]: `fromJson` should take a lambda as well - so that the user doesn't have to parse the json himself to get the attributes.

File: softwareComponents/configuration/include/configuration/serialization.hpp [hunk 4]
```diff
@@ -250,26 +277,30 @@ namespace rofi::configuration::serialization {
...
         Rofibot bot;
 
-        for ( int i = 0; i < j[ "modules" ].size(); i++ ) {
+        for ( unsigned i = 0; i < j[ "modules" ].size(); i++ ) {
```
Comment (Ondraceq) [inline softwareComponents/configuration/include/configuration/serialization.hpp:294]: Use `size_t` when using iterators.

File: softwareComponents/configuration/legacy/configuration/Configuration.h [hunk 5]
```diff
@@ -61,8 +61,8 @@ class Module {
     bool operator==(const Module& other) const;
 
 private:
-    ID id;
     double alpha, beta, gamma;
+    ID id;
```
Comment (Ondraceq) [inline softwareComponents/configuration/legacy/configuration/Configuration.h:65]: Revert this change. (It adds a warning.)

File: softwareComponents/configuration/test/rofibot.cpp [hunk 6]
```diff
@@ -34,13 +34,13 @@ TEST_CASE( "Universal Module Test" ) {
         auto um = UniversalModule( 0, 0_deg, 0_deg, 0_deg );
         CHECK( um.connectors().size() == 6 );
 
-        for ( int i = 0; i < um.connectors().size(); i++ ) {
+        for ( auto i = 0; i < um.connectors().size(); i++ ) {
```
Comment (Ondraceq) [inline softwareComponents/configuration/test/rofibot.cpp:37]: Use `size_t` 0 `auto` here is a fancy way of saying `int` (you will still get `-Wconversion` on the comparison).

File: softwareComponents/configuration/test/rofibot.cpp [hunk 7]
```diff
@@ -174,7 +174,8 @@ TEST_CASE( "Universal Module Test" ) {
         }
 
         SECTION( "radians" ) {
-            auto um = UniversalModule( 0, 0_rad, Angle::rad( -1.57079632679489661923 ), 1.57079632679489661923_rad );
+            auto um = UniversalModule( 0, 0_rad, Angle::rad( -1.57079637 )
+                                               , 1.57079632679489661923_rad );
```
Comment (Ondraceq) [inline softwareComponents/configuration/test/rofibot.cpp:178]: Probably use `Angle::pi / 2` or the `Angle::deg` function.

**Review by Ondraceq — CHANGES_REQUESTED** [review]
I like the changes, here are things to improve.

File: softwareComponents/configuration/include/configuration/serialization.hpp (hunk 1 above)
Comment (yaqwsx) [reply softwareComponents/configuration/include/configuration/serialization.hpp]: Just to be clear: the UB occurs when there are no children; if so, `++j.begin()` is undefined.

File: softwareComponents/configuration/include/configuration/serialization.hpp [hunk 8]
```diff
@@ -233,43 +301,49 @@ namespace rofi::configuration::serialization {
...
             res[ "spaceJoints" ].push_back( j );
         }
 
         return res;
     }
 
-    template<>
-    inline Rofibot fromJSON( const nlohmann::json& j ) {
+    inline nlohmann::json toJSON( const Rofibot& bot ) {
+        return toJSON( bot, []( auto ... ){ return nlohmann::json{}; } );
```
Comment (Ondraceq) [inline softwareComponents/configuration/include/configuration/serialization.hpp:330]: This takes all arguments by value (-> unnecessary copies).

**Review by Ondraceq — CHANGES_REQUESTED** [review]
Every time you use lambda that takes anything and does nothing, take the arguments by universal reference, not by value. (So `[](auto&&...){}` instead of `[](auto...){}`).

**Review by Ondraceq — APPROVED** [review]
